### PR TITLE
[MI-111] Allow user to upload an attachment

### PR DESCRIPTION
**Pull request by miogalang**
Add uploading of attachments to version 2. Allow uploading of attachments and creating tickets with them.

- [x] Update sending to only include options that are available
- [x] Implement with `send_with_options`
- [x] Unit Test
- [x] Test in live

/cc @joseconsador @jwswj @mmolina @atroche 

### References
 - Jira link: https://zendesk.atlassian.net/browse/MI-111

### Risks
 - Ticket attachments might fail (medium)

### CODE REVIEW

**Comment by miogalang** [comment]
Still waiting on https://github.com/aerisweather/GuzzleHttpMock/issues/2 to be able to mock file uploads. The repo doesn't seem to be active so I am crossing my fingers.

**Comment by miogalang** [comment]
I want to test this with authentication before merging. So adding dependency on https://github.com/zendesk/zendesk_api_client_php/pull/115

**Comment by miogalang** [comment]
Finished construction and ready for review

:haircut: @joseconsador @jwswj @atroche @mmolina

**Comment by joseconsador** [comment]
:+1:

File: src/Zendesk/API/Resources/Attachments.php [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        if (! $params['name'] && strrpos($params['file'], '/') > - 1) {
+            $path_array     = explode('/', $params['file']);
+            $file_index     = count($path_array) - 1;
+            $params['name'] = $path_array[$file_index];
+        }
+        if (! $params['name'] && strrpos($params['file'], '/') == false) {
+            $params['name'] = $params['file'];
+        }
+
+        $queryParams = ['filename' => $params ['name']];
```
Comment (jwswj) [inline src/Zendesk/API/Resources/Attachments.php]: an extra space after `$params`?

File: src/Zendesk/API/Resources/Attachments.php [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+     */
+    public function upload(array $params)
+    {
+        if (! $this->hasKeys($params, array('file'))) {
+            throw new MissingParametersException(__METHOD__, array('file'));
+        }
+        if (! file_exists($params['file'])) {
+            throw new CustomException('File ' . $params['file'] . ' could not be found in ' . __METHOD__);
+        }
+        if (! $params['name'] && strrpos($params['file'], '/') > - 1) {
```
Comment (jwswj) [inline src/Zendesk/API/Resources/Attachments.php]: could this be a nested if?
also the logic for getting the name for the path is a bit complicated. is there an easier way to just grab the characters after the last `/`?

File: tests/Zendesk/API/UnitTests/AttachmentsTest.php [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        ];
+
+        $this->client->attachments()->upload($attachmentData);
+
+        $this->assertLastRequestIs(
+            [
+            'method'      => 'POST',
+            'endpoint'    => 'uploads.json',
+            'statusCode'  => 201,
+            'queryParams' => ['filename' => rawurlencode($attachmentData['name'])],
```
Comment (jwswj) [inline tests/Zendesk/API/UnitTests/AttachmentsTest.php]: might want to also test without a name

**Comment by miogalang** [comment]
:haircut:  @jwswj

**Comment by jwswj** [comment]
:+1:

**Pull request merged**